src/main/java/com/ncc/aif/ValidateAIF.java

```diff
@@ -574,4 +592,183 @@ public Resource validateKBAndReturnReport(Model dataToBeValidated, Model union)
+            out.println("Context node: " + queryStats.getContext().toString());
+            out.println("Query Text: " + queryStats.getQueryText().replaceAll("PREFIX.+\n", ""));
+        }
+    }
+
+    /**
+     * A statistics collector for use in profiling TopBraid-based SHACL validation.  Typical usage is to call
+     * {@link #startCollection()} and {@link #endCollection()} to bound statistics collection,
+     * then call {@link #dump(String)} to dump slow query statistics to <filename>-stats.txt.
+     *
+     * This statistics collector progressively dumps slow queries to stdout, which is useful if you suspect
+     * the validation will not complete due to out of memory or other error conditions.
+     */
+    private static class ProgressiveStatsCollector implements ExecStatisticsListener {
+
+        private final int durationThreshold;
+        private final SortedMap<Integer, ExecStatistics> savedStats = new TreeMap<>();
+
+        /**
+         * Creates a statistics collector that progressively dumps queries slower than [threshold] ms to stdout.
+         * @param threshold the threshold definition of a slow query for this statistics collector
+         */
+        ProgressiveStatsCollector(int threshold) {
+            this.durationThreshold = threshold;
+        }
+
+        /**
+         * Start statistics collection.  Clears any previous statistics gathered by TopBraid statistics manager.
+         */
+        void startCollection() {
+            savedStats.clear();
+            ExecStatisticsManager.get().reset();
+            ExecStatisticsManager.get().addListener(this);
+            ExecStatisticsManager.get().setRecording(true);
+        }
+
+        /**
+         * End statistics collection.
+         */
+        void endCollection() {
+            ExecStatisticsManager.get().setRecording(false);
+            ExecStatisticsManager.get().removeListener(this);
+        }
+
+        /**
+         * Receives notification that a TopBraid query statistic has been generated.
+         */
+        public void statisticsUpdated() {
+            final List<ExecStatistics> stats = ExecStatisticsManager.get().getStatistics();
+            final ExecStatistics statistic = stats.get(stats.size() -1);
+            if (statistic.getDuration() > durationThreshold) {
+                savedStats.put(stats.size(), statistic);
+                System.out.println("Dumping slow query #" + stats.size() + "; " + statistic.getDuration() + "ms.");
+                dumpStat(stats.size(), statistic, System.out);
+                System.out.flush(); // Make sure stdout gets displayed even if we eventually run out of memory
+            }
+        }
+
+        /**
+         * Dump all gathered slow query statistics to <basename>-stats.txt, starting with the slowest queries.
+         * @param basename a file basename to determine the profiling output filename
+         */
+        void dump(String basename) {
```

Comment:
It looks like startCollection() endCollection(), and dump() are never used from the ProgressiveStatsCollector class Is ProgressiveStatsCollector even used? I may be missing something here.

Reply:
It's because `ValidateAIF.main( )` instantiates a `StatsCollector`, not a `ProgressiveStatsCollector`.  If you change the instantiation, you should get the opposite situation.

Reply:
I see, would it be useful to expose this to the user through an additional flag so you can alternate between the two without having to change code?

Reply:
I didn't want to complicate the `main()` method any more than it already is, and the profiler is for internal use only, and it's easy to change from one to the other in code.

But I'll give it a shot and see what it does to `main()`.

Reply:
I added a `--p2` switch to enable the `ProgressiveStatsCollector`.